src/web/app/pages-student/student-home-page/student-home-page.component.ts

```diff
@@ -94,67 +96,81 @@ export class StudentHomePageComponent implements OnInit {
     this.hasCoursesLoadingFailed = false;
     this.isCoursesLoading = true;
     this.courses = [];
-    this.courseService.getAllCoursesAsStudent().subscribe((resp: Courses) => {
-      if (!resp.courses.length) {
-        this.isCoursesLoading = false;
-      }
-      for (const course of resp.courses) {
-        this.feedbackSessionsService.getFeedbackSessionsForStudent('student', course.courseId)
-          .subscribe((fss: FeedbackSessions) => {
-            const sortedFss: FeedbackSession[] = this.sortFeedbackSessions(fss);
+    this.courseService.getAllCoursesAsStudent()
+      .pipe(finalize(() => { this.isCoursesLoading = false; }))
+      .subscribe((resp: Courses) => {
+        for (const course of resp.courses) {
+          this.courses.push({
+            course,
+            feedbackSessions: [],
+            isFeedbackSessionsLoading: true,
+            hasFeedbackSessionsLoadingFailed: false,
+          });
+          this.courses.sort((a: StudentCourse, b: StudentCourse) =>
+              ((a.course.courseId > b.course.courseId) ? 1 : -1));
+          this.loadFeedbackSessionsForCourse(course.courseId);
+        }
+      }, (e: ErrorMessageOutput) => {
+        this.hasCoursesLoadingFailed = true;
+        this.statusMessageService.showErrorToast(e.error.message);
+      });
+    }
+
+  /**
+   * Load feedback sessions for a single course.
+   * The course should have been pushed to the this.courses array before this.
+   */
+  loadFeedbackSessionsForCourse(courseId: string): void {
+    // reference to the course within the this.courses array
+    const courseRef = this.courses.find((c) => c.course.courseId === courseId)!;
+    courseRef.isFeedbackSessionsLoading = true;
+    courseRef.hasFeedbackSessionsLoadingFailed = false;
+    this.feedbackSessionsService.getFeedbackSessionsForStudent('student', courseId)
+      .subscribe((fss: FeedbackSessions) => {
+        const sortedFss: FeedbackSession[] = this.sortFeedbackSessions(fss);
+        const studentSessions: StudentSession[] = courseRef.feedbackSessions;
```

Comment:
Is this equivalent to `const studentSessions: StudentSession[] = []`? As the courseRef.feedbackSessions was just initialized

Reply:
No it is not equivalent. This `studentSessions` store the **reference** to the `courseRef.feedbackSessions`. So adding sessions to it will be added to `courseRef.feedbackSessions` and therefore be reflected in the display.